kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,1242 @@
+
+    /// This function checks whether the new app will fit in the bounds dictated by the start address and length provided
+    /// during the setup phase. This function then also computes where in flash the data should be written based on whether
+    /// the call is coming during the app writing phase, or the padding phase.
+    ///
+    /// This function returns the physical address in flash where the write is supposed to happen.
+    fn compute_address(&self, offset: usize, length: usize) -> Result<usize, ErrorCode> {
+        let address = match self.state.get() {
+            Some(State::FirstWrite) | Some(State::AppWrite) => {
+                // Check if there is an overflow while adding length and offset.
+                // If there is no overflow, then we check if the length of the new write block
+                // goes over the total size alloted to the new application.
+                // We also check if the new app is trying to write beyond the bounds of the
+                // flash region allocated to it.
+                // On success, we compute the new address to write the app binary segment.
+                // On failure, we return an Invalid error.
+                match offset.checked_add(length) {
+                    Some(result) => {
+                        if length > self.new_app_length.get() || result > self.new_app_length.get()
+                        {
+                            // this means the app is out of bounds
+                            return Err(ErrorCode::INVAL);
+                        }
+                    }
+                    None => {
+                        return Err(ErrorCode::INVAL);
+                    }
+                }
+                offset + self.new_app_start_addr.get()
+            }
+            // If we are going to write the padding header, we already know where to write in flash,
+            // so we don't have to add the start address
+            Some(State::Setup) | Some(State::Load) | Some(State::PaddingWrite) => offset,
+            // We aren't supposed to be able to write unless we
+            // are in one of the first two write states
+            Some(State::Fail) | Some(State::LastWrite) | None => {
+                return Err(ErrorCode::FAIL);
+            }
+        };
+        Ok(address)
+    }
+
+    /// Compute the physical address where we should write the data and then write it.
+    ///
+    /// This function calls write_done() when the Nonvolatile Driver invokes the callback.
+    fn write(&self, user_buffer: SubSliceMut<'static, u8>, offset: usize) -> Result<(), ErrorCode> {
+        let length = user_buffer.len();
+        let buffer = user_buffer.take(); // for us to perform tbf header validation and write with
+        let physical_address = match self.compute_address(offset, length) {
+            Ok(address) => address,
+            Err(e) => return Err(e),
+        };
+
+        if self.state.get() == Some(State::FirstWrite) {
```

Comment:
I'm concerned about this check. This assumes that the first write has offset=0, but that might not be true. Better to just check based on the provided offset and not an internal state.

Reply:
Also later writes might have offset=0.

Reply:
I don't understand why the first offset would not be 0? The userland app will always try to write from offset 0, correct?

Why would later writes have offset = 0? Do you mean future app load requests? In that case, we reset the internal offset anyway after loading the process, we should be fine (Maybe I should reset the offset after LastWrite, that way we don't have to worry about this in case load never gets called).